### ISSUE TITLE
[Question] Delegation with extension grant for the client credentials flow

### ISSUE THREAD

**Issue by rmja** [IdentityServer/IdentityServer4#775]
I have a setup very similar to the example illustrated in [the extension grants docs](http://docs.identityserver.io/en/release/topics/extension_grants.html), except that instead of having a user initiating the communication to API1, I have an IoT device connecting with the client credentials flow.

The `DelegationGrantValidator` in the docs take a hard dependency on the `sub` claim to be present during the validation, but that is not true for the client credentials flow (The constructor to `GrantValidationResult` throws if `sub` is null).

I think that the behaviour I want is that API1 can, using the original access token, request a new access token with _the same_ claims as those in the original token, with additional scopes, and with a clientId corresponding to API1.

Am I way off track here, or is something along this line possible/secure?

**Comment by leastprivilege** [comment]
We had this issue before - not sure how/if we solved it.

I think you need to new up the result with the default ctor and set the Subject property manually. 

Some claims need to be present though (and this is not obvious) - give it a try.

**Comment by rmja** [comment]
@leastprivilege I will. Would it be correct if the delegation of this access token:
```
{
"client_id": "iot-device",
"device_id": "42", // Custom client claim
"scope": ["api1"]
}
```
would result in this access token:
```
{
"client_id": "api1",
"device_id": "42", // Custom client claim inherited from the original access token.
"scope": ["api2"]
}
```
?

My plan would be to copy all the client claims over to the new access token which are defined in the `Claims` property on the iot-device client configuration. Is that somewhat the right track?

**Comment by leastprivilege** [comment]
Thinking more about it, we might not support this scenario very well.

Give it a try and let me know what happened ;)

**Comment by rmja** [comment]
Ok, I have tried a couple of different approaches without any luck. Here are my findings:

1: The default constructor for `GrantValidationResult` [here](https://github.com/IdentityServer/IdentityServer4/blob/dev/src/IdentityServer4/Validation/Models/GrantValidationResult.cs#L37) throws because it calls another constructor which results in a `NullReferenceException`.

2: I have tried `context.Result = new GrantValidationResult(string.Empty, "delegation", claims: result.Client.Claims)` but the claims included with the claims parameter are not inserted in the resulting access token.

3: I have also tried:
```
context.Result = new GrantValidationResult(string.Empty, "delegation");
var resultClaims = context.Result.Subject.Claims.Where(x => x.Type != JwtClaimTypes.Subject);
                
var id = new ClaimsIdentity(context.Result.Subject.Identity.AuthenticationType);
id.AddClaims(resultClaims.Concat(result.Client.Claims.Where(x => result.Claims.Contains(x, new ClaimComparer(true))).Distinct(new ClaimComparer())));
context.Result.Subject = new ClaimsPrincipal(id);
```
But that throws later in the stack with a "sub claim not present" exception.

Any other ideas?

**Comment by rmja** [comment]
Ok I might be missing something completely. It turns out that the delegated token does not include anything from the original token, such as e.g. profile information, even though the profile scope was in the original token, and also included in the delegation request.

**Comment by leastprivilege** [comment]
OK - https://github.com/IdentityServer/IdentityServer4/blob/dev/src/IdentityServer4/Validation/Models/GrantValidationResult.cs#L37 is clearly a bug.

**Comment by leastprivilege** [comment]
Just by returning claims from the grant validation result - they will not auto-magically show up in the outgoing token.

The principal you create in the validator will be available in your profile service - from there you take what you need to hand it back for token creation. So maybe give your option 2 another try.

**Comment by leastprivilege** [comment]
OK - now I remember again. If you don't set a subject at all, a token for the client only will be created. But we don't have a "legal" way to set the client's claims dynamically right now.

The hacky way would be to modify them straight on the client object on the validated request - but if you are using the in-mem story this will have side effect.

**Comment by leastprivilege** [comment]
See #629

**Comment by leastprivilege** [comment]
This is fixed now in 1.2.0

https://github.com/IdentityServer/IdentityServer4/blob/dev/test/IdentityServer.IntegrationTests/Clients/Setup/DynamicParameterExtensionGrantValidator.cs
